### ISSUE TITLE
Segmentation fault on print()

### ISSUE THREAD

**Issue by thomaswhocares** [GrahamDumpleton/mod_wsgi#650]
Hi,
while running a Django app via wsgi_express i came across a forgotten print() which caused a segmentation fault.
"AH00051: child pid 184567 exit signal Segmentation fault (11), possible coredump in /mod_wsgi-express-80"
After removing said print() the error was gone. 
My stack is:
Ubuntu 20.04
Apache  2.4.41-4
mod_wsgi  4.7.1
Python 3.8.5

I used mod_wsgi express to generate my Apache config and used setup.py to compile/ install wsgi into my virtualenv.

Thank you for your time.

**Comment by GrahamDumpleton** [comment]
What was the ``print()`` statement printing. Where in code was it located? Was it in code that would only be run on process shutdown?

**Comment by thomaswhocares** [comment]
The `print()` was printing a normal string and was located in a regular view. When ever i place `print("test")` somewhere i get a segfault when i open the corresponding page.
I noticed that i also get segfault when i access a page for the first time after a restart.
From reading your documentation i guess this could be a libary version conflict with apache and python on my end.

**Comment by GrahamDumpleton** [comment]
Can you try with:

```
pip install -U https://github.com/GrahamDumpleton/mod_wsgi/archive/develop.zip
```

Is a long shot, but there is an obscure and quite rare memory corruption issue fixed in that version.

**Comment by thomaswhocares** [comment]
The segfault on the first pageload is gone now. But a `print("")` still causes the same problems.

**Comment by GrahamDumpleton** [comment]
Since you are using ``mod_wsgi-express``, try reinstalling but ensure you force that the ``pip`` cache be ignored.

```
pip install -U --no-cache-dir mod_wsgi
```

What might be happening is that you are using a cached binary build that is for a slightly different Apache or Python version and that is somehow causing issues.

Installing the ``develop`` version probably achieved that, but use that ``--no-cache-dir`` option to be sure.

Also see if you can replicate problem with just a simple WSGI hello world program. If not, then what is the set of Python packages your application would be using?

**Comment by thomaswhocares** [comment]
I think i found my problem.
As soon as i `import styleframe`, print("") starts causing stackfaults and all the nice debug info from DEBUG = True gets replaced for any issue with a bland "internal server errror" and a segfault.
Also i was wrong about the develop build fixing some segfaults. Both develop and 4.7.1 react the same.

Below is the issue in a hello world program

`import styleframe;
def application(environ, start_response):
    status = '200 OK'
    output = b'Hello World!'
    print("")
    response_headers = [('Content-type', 'text/plain'),
                        ('Content-Length', str(len(output)))]
    start_response(status, response_headers)

    return [output]`

`[Mon Feb 08 16:28:13.133727 2021] [wsgi:error] [pid 557314:tid 139711086929664] [client 127.0.0.1:43362] Truncated or oversized response headers received from daemon process 'localhost:8000': /tmp/mod_wsgi-localhost:8000:1000/htdocs/`
`[Mon` Feb 08 16:28:13.955197 2021] [core:notice] [pid 557311:tid 139711093972032] AH00051: child pid 557313 exit signal Segmentation fault (11), possible coredump in /tmp/mod_wsgi-localhost:8000:1000`

**Comment by GrahamDumpleton** [comment]
Can only suggest trying to get a stack track as to where it crashes.

* https://modwsgi.readthedocs.io/en/master/user-guides/debugging-techniques.html#debugging-crashes-with-gdb

**Comment by thomaswhocares** [comment]
I found the issue.

import warnings
warnings.warn('...',DeprecationWarning, stacklevel=2)
causes a stackfault for me
It is used in styleframe
https://github.com/DeepSpace2/StyleFrame/blob/master/styleframe/deprecations.py
I just commented out a few lines referencing warnings and now styleframe works for me.

Should I close this issue?

**Comment by GrahamDumpleton** [comment]
I'll see if I can replicate it. Thanks for narrowing it done.

**Comment by GrahamDumpleton** [comment]
If you use a hello world program like:

```
import warnings
warnings.warn('...',DeprecationWarning, stacklevel=2)

def application(environ, start_response):
    status = '200 OK'
    output = b'Hello World!'

    response_headers = [('Content-type', 'text/plain'),
                        ('Content-Length', str(len(output)))]
    start_response(status, response_headers)

    return [output]
```

does it occur.

I suspect that commenting those out has just moved memory around enough that it doesn't occur.

Only a stack trace from the core dump may help to understand it.

**Comment by thomaswhocares** [comment]
Sorry i did not pay enough attention last time.
This exaple crashes. The one from before did indeed not.

I ran this with 
mod_wsgi-express start-server ~/wsgi_hello_world.py

########################################
import warnings

def application(environ, start_response):
    print("") 
    status = '200 OK'
    output = b'Hello World!'

    response_headers = [('Content-type', 'text/plain'),
                        ('Content-Length', str(len(output)))]
    start_response(status, response_headers)

    return [output]

warnings.simplefilter('always', DeprecationWarning)
########################################

I was not able to get a backtrace from the hello world example, but i have one from when i imported styleframe. I dont know how usefull that is.
[apache2.log](https://github.com/GrahamDumpleton/mod_wsgi/files/5963823/apache2.log)
The first webrequest happens at row 253.

Thank you for your time

**Comment by GrahamDumpleton** [comment]
Okay, that replicates issue even on macOS. I will dig into what is going on.

**Comment by GrahamDumpleton** [comment]
So the problem is because when attempting to write log messages to the Apache error log, that is triggering a Python deprecation warning, which causes another attempt to write an error to the Apache error log while still writing the first, which then results in it getting into a never ending loop.

```
442 mod_wsgi-py38.cpython-38-darwin.so	0x000000010b39c6c2 Log_write + 146
443 mod_wsgi-py38.cpython-38-darwin.so	0x000000010b39c67b Log_write + 75
444 org.python.python             	0x000000010b3e9ebe method_vectorcall_VARARGS + 270
445 org.python.python             	0x000000010b3e475a object_vacall + 352
446 org.python.python             	0x000000010b3e45ba PyObject_CallMethodObjArgs + 203
447 org.python.python             	0x000000010b5054db _textiowrapper_writeflush + 364
448 org.python.python             	0x000000010b506091 _io_TextIOWrapper_write + 905
449 org.python.python             	0x000000010b3ea679 method_vectorcall_O + 244
450 org.python.python             	0x000000010b48367d call_function + 346
451 org.python.python             	0x000000010b48023b _PyEval_EvalFrameDefault + 29833
452 org.python.python             	0x000000010b3e32c2 function_code_fastcall + 106
453 org.python.python             	0x000000010b48367d call_function + 346
454 org.python.python             	0x000000010b480314 _PyEval_EvalFrameDefault + 30050
455 org.python.python             	0x000000010b3e32c2 function_code_fastcall + 106
456 org.python.python             	0x000000010b3e475a object_vacall + 352
457 org.python.python             	0x000000010b3e4965 PyObject_CallFunctionObjArgs + 136
458 org.python.python             	0x000000010b455d8f warn_explicit + 2132
459 org.python.python             	0x000000010b456a82 do_warn + 363
460 org.python.python             	0x000000010b4554c0 warn_unicode + 35
461 org.python.python             	0x000000010b455477 PyErr_WarnEx + 48
462 org.python.python             	0x000000010b497f4c convertitem + 4081
463 org.python.python             	0x000000010b4946fd vgetargs1_impl + 552
464 org.python.python             	0x000000010b494371 PyArg_ParseTuple + 126
465 mod_wsgi-py38.cpython-38-darwin.so	0x000000010b39c6c2 Log_write + 146
466 mod_wsgi-py38.cpython-38-darwin.so	0x000000010b39c67b Log_write + 75
467 org.python.python             	0x000000010b3e9ebe method_vectorcall_VARARGS + 270
468 org.python.python             	0x000000010b3e475a object_vacall + 352
469 org.python.python             	0x000000010b3e45ba PyObject_CallMethodObjArgs + 203
470 org.python.python             	0x000000010b5054db _textiowrapper_writeflush + 364
471 org.python.python             	0x000000010b506091 _io_TextIOWrapper_write + 905
472 org.python.python             	0x000000010b3ea679 method_vectorcall_O + 244
473 org.python.python             	0x000000010b48367d call_function + 346
474 org.python.python             	0x000000010b48023b _PyEval_EvalFrameDefault + 29833
475 org.python.python             	0x000000010b3e32c2 function_code_fastcall + 106
476 org.python.python             	0x000000010b48367d call_function + 346
477 org.python.python             	0x000000010b480314 _PyEval_EvalFrameDefault + 30050
478 org.python.python             	0x000000010b3e32c2 function_code_fastcall + 106
479 org.python.python             	0x000000010b3e475a object_vacall + 352
480 org.python.python             	0x000000010b3e4965 PyObject_CallFunctionObjArgs + 136
481 org.python.python             	0x000000010b455d8f warn_explicit + 2132
482 org.python.python             	0x000000010b456a82 do_warn + 363
483 org.python.python             	0x000000010b4554c0 warn_unicode + 35
484 org.python.python             	0x000000010b455477 PyErr_WarnEx + 48
485 org.python.python             	0x000000010b497f4c convertitem + 4081
486 org.python.python             	0x000000010b4946fd vgetargs1_impl + 552
487 org.python.python             	0x000000010b494371 PyArg_ParseTuple + 126
488 mod_wsgi-py38.cpython-38-darwin.so	0x000000010b39c6c2 Log_write + 146
489 mod_wsgi-py38.cpython-38-darwin.so	0x000000010b39c67b Log_write + 75
490 org.python.python             	0x000000010b3e9ebe method_vectorcall_VARARGS + 270
491 org.python.python             	0x000000010b3e475a object_vacall + 352
492 org.python.python             	0x000000010b3e45ba PyObject_CallMethodObjArgs + 203
493 org.python.python             	0x000000010b5054db _textiowrapper_writeflush + 364
494 org.python.python             	0x000000010b506091 _io_TextIOWrapper_write + 905
495 org.python.python             	0x000000010b3ea679 method_vectorcall_O + 244
496 org.python.python             	0x000000010b48367d call_function + 346
497 org.python.python             	0x000000010b48023b _PyEval_EvalFrameDefault + 29833
498 org.python.python             	0x000000010b3e32c2 function_code_fastcall + 106
499 org.python.python             	0x000000010b48367d call_function + 346
500 org.python.python             	0x000000010b480314 _PyEval_EvalFrameDefault + 30050
501 org.python.python             	0x000000010b3e32c2 function_code_fastcall + 106
502 org.python.python             	0x000000010b3e475a object_vacall + 352
503 org.python.python             	0x000000010b3e4965 PyObject_CallFunctionObjArgs + 136
504 org.python.python             	0x000000010b455d8f warn_explicit + 2132
505 org.python.python             	0x000000010b456a82 do_warn + 363
506 org.python.python             	0x000000010b4554c0 warn_unicode + 35
507 org.python.python             	0x000000010b455477 PyErr_WarnEx + 48
508 org.python.python             	0x000000010b497f4c convertitem + 4081
509 org.python.python             	0x000000010b4946fd vgetargs1_impl + 552
510 org.python.python             	0x000000010b494371 PyArg_ParseTuple + 126
511 mod_wsgi-py38.cpython-38-darwin.so	0x000000010b39c6c2 Log_write + 146
```

Eventually it crashes as exhausts the stack or some other issue.

Not sure how to fix the issue right now.

**Comment by GrahamDumpleton** [comment]
I think it may relate to:

* https://github.com/GrahamDumpleton/mod_wsgi/issues/645

If I fix address that issue, then I am guessing the possibility of a deprecation warning in that code will vanish.

**Comment by thomaswhocares** [comment]
Thats good to know.
Thank you for looking into it.

**Comment by GrahamDumpleton** [comment]
Try with latest ``develop`` version:

```
pip install -U --no-cache-dir https://github.com/GrahamDumpleton/mod_wsgi/archive/develop.zip
```

**Comment by thomaswhocares** [comment]
Thank you it looks like all is working perfectly now even with the styleframe module.
What i also like is that the usage of println() now shows as an nice error inside the apache error log.
`[Sat Feb 13 11:53:40.010267 2021] [wsgi:error] [pid 1016494:tid 139653609867008] [remote 172.23.97.249:27044] test`
Is this new or just for the time being ?

**Comment by GrahamDumpleton** [comment]
Using ``print()``, or otherwise writing to ``sys.stdout`` or ``sys.stderr`` always sent output to the Apache error log.

**Comment by aggieNick02** [comment]
Just FYI, this bit me as well, upgrading from ubuntu 18.04 to 20.04 with a django site. I played whack-a-mole for a bit knocking out calls to logging, etc, before I realized the cause was any print after promoting DeprecationWarning to raise an exception. It wasn't until I changed DeprecationWarning to print that I saw the actual warning, which was enough to google and end up on this issue.

I raise DeprecationWarning to error because printed warnings are often lost in noisy log files, and do it for all DeprecationWarning (instead of those just called from __main__) because much of our code is not called from __main__.

Previously this has led to exceptions from 3rd party code that were easy enough to see and suppress based on module name. But this one with its failure mode took a while to recognize the cause.
